From 9400941ee47c455aeb89342eea6a8ee2b5c8d413 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20MBAPE?= <rose.martin-8og3raxw@yopmail.com>
Date: Sun, 11 Oct 2015 12:28:32 +0200
Subject: [PATCH] fix helpers.coffee inaccurate module not found

---
 src/helpers.coffee | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/helpers.coffee b/src/helpers.coffee
index 238ba5f..2a033d1 100644
--- a/src/helpers.coffee
+++ b/src/helpers.coffee
@@ -433,7 +433,7 @@ exports.loadConfig = (configPath = 'brunch-config', options = {}, callback) ->
     unless config?.files
       throw new Error 'Brunch config must have "files" property'
   catch error
-    if configPath is 'brunch-config' and error.code is 'MODULE_NOT_FOUND'
+    if configPath is 'brunch-config' and error.code is 'MODULE_NOT_FOUND' && /\bbrunch-config\b/.test error.message
       # start to warn about deprecation of 'config' with 1.8 release
       # seamless and silent fallback until then
       return exports.loadConfig 'config', options, callback
-- 
2.5.3.windows.1

