From d2505cd474437ec81e6f163f4c0268a6a464a923 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20MBAPE?= <rose.martin-8og3raxw@yopmail.com>
Date: Tue, 31 Jan 2017 17:20:07 +0100
Subject: [PATCH] add config compiler feature

---
 lib/utils/plugins.js | 17 ++++++++++------
 1 file changed, 11 insertion(+), 6 deletion(-)

diff --git a/lib/utils/plugins.js b/lib/utils/plugins.js
index 0cce810..592930d 100644
--- a/lib/utils/plugins.js
+++ b/lib/utils/plugins.js
@@ -72,14 +72,21 @@ const plugins = (config, craDeps) => {
     return true;
   });
 
+  const addPlugin = (plugins, Plugin, name) => {
+    if (Plugin && Plugin.prototype && Plugin.prototype.brunchPlugin) {
+      const plugin = new Plugin(config);
+      if ('string' !== typeof plugin.brunchPluginName) {
+        plugin.brunchPluginName = 'string' === typeof name ? name : Plugin.brunchPluginName;
+      }
+      plugins.push(adapter(plugin));
+    }
+    return plugins;
+  };
+
   const plugins = deps.reduce((plugins, name) => {
     try {
       const Plugin = require(sysPath.join(npmPath, name));
-      if (Plugin && Plugin.prototype && Plugin.prototype.brunchPlugin) {
-        const plugin = new Plugin(config);
-        plugin.brunchPluginName = name;
-        plugins.push(adapter(plugin));
-      }
+      addPlugin(plugins, Plugin, name);
     } catch (error) {
       if (error.code === 'MODULE_NOT_FOUND' && name in pkg.dependencies) {
         throw new BrunchError('RUN_NPM_INSTALL', {error});
@@ -87,7 +92,7 @@ const plugins = (config, craDeps) => {
       logger.warn(`Loading of ${name} failed due to`, error);
     }
     return plugins;
-  }, [])
+  }, Array.isArray(config.compilers) ? config.compilers.reduce(addPlugin, []) : [])
   .filter(plugin => {
     // Does the user's config say this plugin should definitely be used?
     if (on.includes(plugin.brunchPluginName)) return true;

-- 
2.10.0.windows.1

