diff --git a/lib/utils/modules.js b/lib/utils/modules.js
index 85ad6ab..cffaae3 100644
--- a/lib/utils/modules.js
+++ b/lib/utils/modules.js
@@ -45,8 +45,8 @@ exports.normalizeWrapper = (wrapper, nameCleaner) => {
   const wrapperFn = normalizeResult(getWrapperFn(wrapper));
 
   return (path, compiled) => {
-    const name = sysPath.normalize(path).replace(/^(\.\.\/)+/, '');
-    return wrapperFn(nameCleaner(name), compiled);
+    const name = sysPath.normalize(path);
+    return wrapperFn(nameCleaner(name).replace(/^(\.\.\/)+/, ''), compiled);
   };
 };
 
