diff --git a/lib/watch.js b/lib/watch.js
index 6c674f9..0458376 100644
--- a/lib/watch.js
+++ b/lib/watch.js
@@ -110,6 +110,25 @@ class BrunchWatcher {
     }
 
     application.loadConfig(persistent, options)
+      .then(cfg => {
+        return new Promise(resolve => {
+          if (typeof cfg.initialize === 'function') {
+            const initTimeout = setTimeout(function() {
+              logger.warn('initialize is taking a long time to start');
+              return logger.warn('**don\'t forget to invoke callback()**');
+            }, cfg.timeout || 5000);
+
+            const callback = () => {
+              clearTimeout(initTimeout);
+              resolve(cfg);
+            };
+
+            cfg.initialize(cfg, callback);
+          } else {
+            resolve(cfg);
+          }
+        });
+      })
       .then(cfg => {
         this.config = cfg;
         if (options.jobs > 1) {
@@ -136,6 +155,10 @@ class BrunchWatcher {
 
         return Promise.all(this.hooks.preCompile()).then(() => {
           this.initWatcher(watchedPaths);
+          const cfg = this.config;
+          if (typeof cfg.onwatch === 'function') {
+            cfg.onwatch(this.watcher, this);
+          }
           this.initCompilation();
         });
       })
