diff --git a/lib/utils/config.js b/lib/utils/config.js
index 0f5a105..09d5272 100644
--- a/lib/utils/config.js
+++ b/lib/utils/config.js
@@ -6,7 +6,7 @@ const checkDeps = require('check-dependencies');
 const sysPath = require('universal-path');
 const anymatch = require('anymatch');
 const logger = require('loggy');
-const readComponents = require('read-components');
+const readComponents = require('./read-components');
 const debug = require('debug')('brunch:config');
 const helpers = require('./helpers');
 const deppack = require('deppack'); // isNpm, loadInit
diff --git a/lib/utils/helpers.js b/lib/utils/helpers.js
index f3ac53a..e79c486 100644
--- a/lib/utils/helpers.js
+++ b/lib/utils/helpers.js
@@ -3,6 +3,8 @@ const basename = require('universal-path').basename;
 const promisify = require('micro-promisify');
 const fslstat = promisify(require('fs').lstat);
 const fsaccess = promisify(require('fs').access);
+const anymatch = require('anymatch');
+const explore = require('fs-explorer').explore;
 
 // Single-level flatten.
 const flatten = array => [].concat.apply([], array);
@@ -199,7 +201,35 @@ const deepFreeze = (object, except) => {
 };
 
 const fsExists = path => {
-  return fsaccess(path).then(() => true, () => false);
+  // http://www.regular-expressions.info/characters.html#special
+  if (/[\^$|?*+()[\]{}]/.test(path)) {
+    return new Promise((resolve, reject) => {
+      // matchers are relative to component directory
+      // path = <basedir>/<component>/relative
+      // componentDir = <basedir>/<component>
+      const componentDir = path.replace(/([^/\\]+[/\\][^/\\]+)[/\\].*/g, '$1');
+
+      const matcher = anymatch([path]);
+      const files = [];
+      explore(componentDir, (path, stats, next) => {
+        // replace(/[\\]/g, '/') to avoid path sep OS differencies
+        if (matcher(path.replace(/\\/g, '/'))) {
+          files.push(path);
+        }
+        next();
+      }, {followSymlink: true}, err => {
+        if (err && err.code === 'ENOENT') {
+          err = null;
+        }
+        if (err) {
+          reject(err);
+        } else {
+          resolve(files);
+        }
+      });
+    });
+  }
+  return fsaccess(path).then(() => path, () => false);
 };
 
 const isSymlink = path => {
diff --git a/lib/watch.js b/lib/watch.js
index 892bfb3..6c674f9 100644
--- a/lib/watch.js
+++ b/lib/watch.js
@@ -47,9 +47,18 @@ const mergeHooks = (plugins, config) => {
 };
 
 const filterNonExistentPaths = paths => {
-  return Promise.all(paths.map(helpers.fsExists)).then(values => {
+  return Promise.all(paths.map(helpers.fsExists)).then(files => {
     // watched files
-    return paths.filter((path, index) => values[index]);
+    const watchedFiles = [];
+    for (let i = 0, len = files.length; i < len; i++) {
+      const file = files[i];
+      if (Array.isArray(file)) {
+        watchedFiles.push.apply(watchedFiles, file);
+      } else if (file) {
+        watchedFiles.push(file);
+      }
+    }
+    return watchedFiles;
   });
 };
 
diff --git a/package.json b/package.json
index fcd4f88..daa2b23 100644
--- a/package.json
+++ b/package.json
@@ -59,6 +59,7 @@
     "deppack": "~0.7",
     "deps-install": "~0.1",
     "fcache": "~0.3",
+    "fs-explorer": "^1.1.0",
     "init-skeleton": "~1.0",
     "loggy": "~1.0.2",
     "micro-es7-shim": "^0.1",
