--- node_modules/log4js/lib/log4js.js.orig	2016-02-12 16:42:33.461550500 +0100
+++ node_modules/log4js/lib/log4js.new.js	2016-02-12 16:45:57.437899400 +0100
@@ -428,6 +428,12 @@
   // not being able to be drained because of run-away log writes.
   loggerModule.disableAllLogWrites();
 
+  // clear timer
+  if (configState.timerId) {
+    clearInterval(configState.timerId);
+    delete configState.timerId;
+  }
+
   // Call each of the shutdown functions in parallel
   var completed = 0;
   var error;
