--- a/node_modules/log4js/lib/log4js.js
+++ b/node_modules/log4js/lib/log4js.js
@@ -454,6 +454,12 @@
   // not being able to be drained because of run-away log writes.
   loggerModule.disableAllLogWrites();
 
+  // clear timer
+  if (configState.timerId) {
+    clearInterval(configState.timerId);
+    delete configState.timerId;
+  }
+
   // Call each of the shutdown functions in parallel
   var completed = 0;
   var error;
